### ISSUE TITLE
Transactions that revert with binary data reasons do not return the data -- this breaks gnosis safe relaying

### ISSUE THREAD

**Issue by habdelra** [NethermindEth/nethermind#3207]
When a transaction fails a solidity `require()` that has a binary data "reason", the response from nethermind does not return the binary data. rather it looks like nethermind assumes all revert reasons are strings and is trying to cast the binary data to unicode. This is very problematic for using gnosis safes. Specifically the gnosis safe relay server relies on intentional reverts in order to communicate gas estimate information in the revert reason code as binary data. You can see an example of this code here:
https://github.com/gnosis/safe-contracts/blob/main/contracts/GnosisSafe.sol#L306-L327 and here https://github.com/gnosis/safe-contracts/blob/main/contracts/common/StorageAccessible.sol#L44. When the gnosis relay server attempts to intentionally trigger a revert and read the revert reason as binary data using a nethermind node, it throws an exception because the revert reason completely obscures the original binary data that was to accompany the revert reason. This was initially filed as a gnosis bug here https://github.com/gnosis/gnosis-py/issues/26, and after investigation it was realized this is actually a nethermind issue. 

Now that nethermind is _the_ recommended ethereum node for xDai, POA, and Sokol, this is a huge problem as we won't be able to reliably run gnosis safe transactions in those networks since they will be running nethermind.

**To Reproduce**
To reproduce this issue, post the following message (this triggers the intentional revert):
POST `https://rpc.xdaichain.com/`
```
{
    "jsonrpc": "2.0",
    "method": "eth_call",
    "params": [
        {
            "value": 0,
            "chainId": 100,
            "from": "0x9a0bbbbd3789f184CA88f2F6A40F42406cb842AC",
            "gasPrice": 0,
            "to": "0x9a0bbbbd3789f184CA88f2F6A40F42406cb842AC",
            "data": "0xc4ca3a9c000000000000000000000000333d72f54cf17ba7a1971e7c7e9acb91c8b709d20000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000008000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000004a2e6204500000000000000000000000000000000000000000000000000000000"
        }
    ],
    "id": 1
}
```
This returns the response:
```
{"jsonrpc":"2.0","error":{"code":-32015,"message":"VM execution error.","data":"revert: \u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000?{"},"id":1}
```

You can confirm that this node is actually a nethermind node by sending the following message:
POST `https://rpc.xdaichain.com/`
```
{"jsonrpc":"2.0","method":"web3_clientVersion","params":[],"id":1}
```
This returns the response:
```
{
    "jsonrpc": "2.0",
    "result": "Nethermind/v1.10.73-0-b8ab96510-20210625/X64-Linux/5.0.7",
    "id": 1
}
```

**Expected behavior**
The response should look something like this:
```
{
    "jsonrpc": "2.0",
    "error": {
        "code": -32015,
        "message": "VM execution error.",
        "data": "Reverted 0x08c379a0000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000002123"
    },
    "id": 1
}
```
Where the binary data is actually returned in the error data. The gnosis relay server is expecting to find binary data here that it can parse to use for calculating gas estimates.

**Comment by tkstanczak** [comment]
acknowledge, triaging

**Comment by chris-shyft** [comment]
guys I'm having a hard time rectifying something in my head -

1. assume POA sealers, common compatible genesis file

2. assume 50% openethereum, 50% nethermind implementations

3. gnosis 1.3.0 deployed to the network

4. someone creates a gnosis safe, works on both clients

5. someone triggers a add_owner command in safe-cli

6. 50% of the time, it will be approved, 50% it will not (depends on rpc/sealing). 

7. Sometimes it might say like it'll work and perform the transaction if the RPC endpoint is an OE one.

8. in the case where the tx passes by #7 scenario here, the block explorer will show a revert error.

9. in the case where the OE rpc endpoint allows it though, *and* the OE sealer finds the tx and seals it...


-----
question A:
what happens after #8 occurs?
the round robin has 50% of the network as OE sealers, this will cause the tx to pass on their checks, which will be regarded as a different block hash, so they won't accept the sealed block?

question B:
what happens after #9 occurs?
do the rest of the NM sealers at the next block reject the transaction? 
-------


this would seem to me, if the answers above to questions A & B are as I would expect, this type of transaction could be weaponized and crack networks in half (or worse).

can anyone elucidate as to the scenario outcomes above?

**Comment by chris-shyft** [comment]
and ofc what happens with an independent:

scenario C: NM node starts to re-synch blocks from block 0 w/ a full archive sync
scenario D: OE node  starts to re-synch blocks from block 0 w/ a full archive sync

**Comment by habdelra** [comment]
@chris-shyft this sounds different than what was originally posted. Could you please file as separate issue.

**Comment by tkstanczak** [comment]
@habdelra let us bump priority on this one, hope we can pick it up tomorrow

**Comment by tkstanczak** [comment]
@marcindsobczak @MarekM25 -> something that you can arrange with a team for a fix early this week?

**Comment by igorbarinov** [comment]
We are also interested in deploying of the fix
